Fixed #31117 -- Isolated backends.base.test_creation.TestDbCreationTests

matthijskooijman opened the pull request:
(Commit replaced with different fix, see comments below)

~~This lets TestDbCreationTests properly clean up by not only calling `create_test_db`, but also `destroy_test_db`.~~

~~Since we cannot call `destroy_test_db` on the 'default' database, which is used by other tests, and since calling `create_test_db` on a database that is already created is fragile in any case, this also changes the test to use the 'unused' database that is introduced in this PR.~~

~~Additionally, this changes the restoration of the connection settings to restore just the `MIGRATE` setting, rather than the entire dict, since that prevents issues when the same dict is referenced from multiple places (as detailed in #31117).~~

~~Before this PR can be merged, https://code.djangoproject.com/ticket/31055 must be fixed (work underway in #12201). Until then, this PR contains a hack to disable checking entirely.~~

~~I am not entirely happy with the name of the new database. I called it 'unused', since it is *normally* unused by testcases (e.g. not used directly), but can be used by testcases that need to go through the full create/destroy cycle. Any suggestions?~~

~~Also, when this database will be used for multiple tests (e.g. I also need it for #12166), I'm afraid there might be concurrency issues when both tests end up running at the same time. Is there any standard way to deal with this? Maybe simply add a lock for using this third database?~~

charettes commented on django/test/runner.py:
```diff
@@ -633,7 +633,8 @@ def get_test_runner_kwargs(self):
         # Checks are run after database creation since some checks require
         # database access.
-        call_command('check', verbosity=self.verbosity)
+        #call_command('check', verbosity=self.verbosity)
```
FWIW this will likely not be necessary if https://github.com/django/django/pull/12201 lands as it will allow passing `databases=self.databases` to the command and disable running checks on unrelated ones.

matthijskooijman replied:
Yeah, this was just a temporary hack to at least make things work until that PR is merged.

matthijskooijman commented:
I did some experiments to figure out why this bug did not trigger in Jenkins, but now force pushed those experiments away and continued them in #12248. Compared to my first push, I fixed some flake8 errors, added another fix for a similar issue in `backends.sqlite.test_creation` and made the workaround for #12201 a bit more thorough. With that, all tests seem to be working again now on Jenkins.

matthijskooijman commented:
(cross-post from https://code.djangoproject.com/ticket/31117)

This PR can now again run the testsuite completely successfully. Only things open are:
 - How to prevent concurrency issues accessing this new database when multiple testcases use it? Normally, this is handled by making the testrunner make clones of the database, but I can't see how that would work here. Any suggestions?
 - What name to use for this new database? 'unused' does not seem too great.
 - The Jenkins database configuration needs to be changed to configure this new database. Is there anything I can do for that?
 - https://github.com/django/django/pull/12201 should be merged first.

felixxm commented:
@matthijskooijman Thanks for this patch, however IMO it's too invasive, we can get a proper isolation by creating a test connection more carefully (as a copy of `connections[DEFAULT_DB_ALIAS]`), see #12256. We should be able to use the same approach in `tests/backends/base/test_creation.py`.

matthijskooijman commented:
> Thanks for this patch, however IMO it's too invasive, 

Probably, yeah. It seemed easy since I was experimenting with an extra db already for #12166, but was not so easy as I had hoped. For that other PR, I might be able to reduce the tests and remove the need for an extra db as well.

As for improving isolation for this test, I think your #12256 looks good, and the same approach will probably work for `TestDbCreationTests` as well. Using a new connection object, it is probably safe enough to call both `create_test_db` and `destroy_test_db`. One point of attention is that `create_test_db` modifies `django.conf.settings.DATABASES` as well, which is not isolated by creating a a new connection (which is why I would consider calling `destroy_test_db` as well).


Regardless, would you maybe consider merging 283eac71afd08e1d01e78a62c6f134d359d758f7 anyway, even when no additional db will be used? Most parts of that commit simply make testcases more self-contained and less reliant on the test configuration, without complicating the tests in any way (some parts of the commit are a bit more elaborate and might be better to drop if not directly used).

felixxm commented:
> Regardless, would you maybe consider merging [283eac7](https://github.com/django/django/commit/283eac71afd08e1d01e78a62c6f134d359d758f7) anyway, even when no additional db will be used? Most parts of that commit simply make testcases more self-contained and less reliant on the test configuration, without complicating the tests in any way (some parts of the commit are a bit more elaborate and might be better to drop if not directly used).

It makes sense at first glance, but let's do this separately because it's not strictly related with an isolation issue.

matthijskooijman commented:
@felixxm, I updated this PR, fixing the test in a more simple way like you suggested. I've tested this in a few combinations locally, and also intentionally broke the code to see if the testcase would still catch that, which it did.

> > Regardless, would you maybe consider merging 283eac7 anyway
> It makes sense at first glance, but let's do this separately because it's not strictly related with an isolation issue.

I've created PR #12342 for this.

felixxm commented:
@matthijskooijman Thanks for updates :+1: I moved `TestDbSignatureTests.get_connection_copy()` test hook to a module level and used it in your fix.

matthijskooijman commented:
I was just looking at the changes you made, looks good!
